### ISSUE TITLE
Catching infinite instantiations more intelligently

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#13692]
tl;dr: It seems as though we should be able to detect infinite instantiations in a more robust manner than just counting the number of instantiations of a function made.  But how?

The compiler currently guards against infinite instantiations, which can cause an infinite loop in the compiler (until it runs out of memory) like the following:

```chapel
proc foo(param i : int) {
  writeln("in foo ", i);
  foo(i + 1);
}

foo(1);
```

However, the way it currently does so is simply by counting the number of instantiations of `foo()` and assuming that we're in the infinite instantiation case once it exceeds some limit (currently set to 256).

This is a somewhat weak check, though, since it can cause programs with reasonable / legal instantiations (albeit many of them) to not compile until the limit is raised to some other semi-arbitrary limit.  For example, Arkouda currently runs into this limit with the `isArray()` and `canResolveMethod()` queries because it's a big enough code that we create a lot of instantiations of them.

Ultimately, I think that the compiler would benefit from doing fewer instantiations in terms of compile-time and memory and think that's something we should pursue (see discussion on issue #10851 and particularly [this comment](https://github.com/chapel-lang/chapel/issues/10851#issuecomment-519715526) relating to this Arkouda investigation).

However, in the meantime, it seems like we could be smarter about detecting infinite instantiation loops rather than just counting the number of instantiations.  For example, given the definition of `canResolveMethod()`, it's clear as a human to see that it can't result in infinite instantiations because it isn't recursive.  What would it take to do this in the compiler?  Similarly for `isArray()`, though determining that case requires knowing that `isArrayValue()` doesn't call recursively back to `isArray()`, so it's slightly more challenging.

**Comment by bradcray** [comment]
I was initially thinking that this was just a matter of detecting recursive functions in the compiler, but it's arguably slightly more complicated than that since the recursion is w.r.t. the generic function, not the concrete instantiations (which are not technically recursive).  So it seems _like_ detecting recursion, but also slightly different.

My main thought here is whether we can store a collection of generic function symbols that we're currently working on instantiating and look for duplicate entries as we do so (which might suggest recursion; though not necessarily infinite recursion, of course... so we'd probably still want a limit for that).  And then to clear out this collection whenever we're done with a particular instantiation.  Thus, for a case like `isArray()` and `isArrayValue()`, we'd add them to the collection, finish off their instantiations, and then clear them out (no recursion detected, and we're back to a blank slate the next time we see them which will also be safe).

My main questions are where would such code be inserted?  And do we really need a stack of these collections since the first routine called in a user program could be generic and wouldn't be considered non-recursive / done until the whole program was resolved?  (meaning the collection would persist throughout the whole compilation)

**Comment by bradcray** [comment]
I think the solution is actually much easier than I was imagining.  Currently, we increment a per-generic-function counter each time we instantiate it, and if we surpass the max, we give the error.  But maybe the fix isn't as complex as I was hypothesizing above; maybe we simply need to decrement the counter once the resolution of the function is complete.  That would catch the infinite recursion case (since it would get bumped up to 256 as it was recursively instantiated) while letting the isArray() case go through (since it would be bumped up to 1 and then down to 0 repeatedly each time it was considered).

So then the only question is where to bump the counter down... maybe where we add FLAG_RESOLVED to the function?

**Comment by bradcray** [comment]
> maybe where we add FLAG_RESOLVED to the function?

Oops, no.  It looks like we add that when we start resolving a function, not when we complete it.

@mppf / @vasslitvinov : Do either of you have suggestions here?  (if the idea even makes sense)

**Comment by mppf** [comment]
@bradcray - sure, I'd recommend putting it here:

https://github.com/chapel-lang/chapel/blob/693b06683244b4a000f213bcecd441534d4c7b6e/compiler/resolution/resolveFunction.cpp#L508

(in resolveFunction at the end of the body of the `if (fn->isResolved() == false) {` block).

**Comment by bradcray** [comment]
That seems to be working pretty well, thanks!

**Comment by bradcray** [comment]
PR #13693 has a proposed solution to the problem that's passing tests specifically about instantiation.  I now need to make sure it doesn't break anything else (which would probably only happen if we had unbalanced increment/decrement counts because I put a pretty string INT_FATAL() in to check for that).  Thanks for the assist, @mppf!
